src/web/components/App.tsx

```diff
@@ -158,6 +163,11 @@ export const App = ({ CAPI, NAV, ophanRecord }: Props) => {
 	>();
 
 	const pageViewId = window.guardian?.config?.ophan?.pageViewId;
+	const [browserId, setBrowserId] = useState<string>();
+	useEffect(() => {
+		// TODO: can the browserId actually be null?
+		setBrowserId(getCookie('bwid') ?? undefined);
+	}, []);
```

Comment:
This `useEffect` means this code only runs after rendering has completed, and only once (because of the empty deps array). No idea if this is the case or not but do we need that? Could this code run straight away, before render?

Reply:
Ah, yes, I think that would be better, I’m not super familiar with the React lifecycle here!

Reply:
> Any idea how we’d do that? Just use a `const` one level higher?

So yes, using a const would be the alternative. However, the down side to doing _this_ is `document.cookie` is synchronous and has performance issues so this code will slow down hydration. Sorry, I didn't think that one through before making my original comment!

It's a trade off. Some information is critical and worth blocking the lifecycle for because it's required to make decisions but in other situations it's better to bring the information in asynchronously (after the App code has already been run).

### However
There's another factor to consider here. By adding state to `App`, changing that state will cause the whole file to rerender which could have unintended consequences in other areas. This is a structural problem with how `App` has been built, it advertises itself as the place where things like this should go but actually it might not be. To avoid rerenders in other areas you can drop the state down into a kind of wrapper or container component, using composition to control rendering.

Reply:
So, would you suggest leveraging `useOnce`? This way this is only ever set once.

Should this also contain the lookup for checking if a user is signed-in? I’m looking at `!!getCookie('GU_U')` on L200.